File: Codes/1465-maximum-area-of-a-piece-of-cake-after-horizontal-and-vertical-cuts.cpp
/*
 * @Author: LetMeFly
 * @Date: 2023-10-27 12:24:04
 * @LastEditors: LetMeFly
 * @LastEditTime: 2023-10-27 12:29:31
 */
#ifdef _WIN32
#include "_[1,2]toVector.h"
#endif

class Solution {
private:
    long long getMax(int l, vector<int>& v) {
        sort(v.begin(), v.end());
        int ans= 0;
        for (int i = 1; i < v.size(); i++) {
            ans = max(ans, v[i] -  v[i - 1]);
        }
        return max(ans, max(v[0], l - v[v.size() - 1]));
    }

public:
    int maxArea(int h, int w, vector<int>& horizontalCuts, vector<int>& verticalCuts) {
        return getMax(h, horizontalCuts) *  getMax(w, verticalCuts) % 1000000007;
    }
};

File: Codes/1465-maximum-area-of-a-piece-of-cake-after-horizontal-and-vertical-cuts.py
'''
Author: LetMeFly
Date: 2023-10-27 12:30:02
LastEditors: LetMeFly
LastEditTime: 2023-10-27 12:33:31
'''
from typing import List

class Solution:
    def getMax(self, l: int, v: List[int]) -> int:
        v.sort()
        ans = v[0]
        for i in range(1, len(v)):
            ans = max(ans, v[i] - v[i - 1])
        return max(ans, l - v[-1])

    def maxArea(self, h: int, w: int, horizontalCuts: List[int], verticalCuts: List[int]) -> int:
        return self.getMax(h, horizontalCuts) * self.getMax(w, verticalCuts) % 1000000007